book/src/leader-validator-transition.md

```diff
@@ -0,0 +1,157 @@
+# Leader and Validator Transition
+
+A fullnode operates in two modes, leader and validator.  The modes overlap in
+code and in function, but have different behaviors.  The goal for this design is
+to allow the two modes to transition between states while reusing the computed
+bank state.
+
+
+## Validator
+
+A validator operates on may different concurrent forks of the bank state until
+it can prove a PoH record at a height that is the scheduled slot to be a leader.
```

Comment:
> prove a PoH record at a height that is the scheduled slot to be a leader

Awkward wording and not very precise. This both suggests that the TVU won't operator on concurrent forks during its slot and that it doesn't return to validating after its slot is over.

Reply:
I worry about this, I think the TVU needs to always be running, that the TPU's fork should be seen as "currently best", but always one of many that have to concurrently be followed.

Reply:
@rob-solana which part of the TVU needs to be running in your mind?  What i am proposing is that EntryTree is still running outside of the loop

Reply:
1. retransmit_stage needs to keep running
2. something like replay_stage needs to keep running for other forks
3. blocktree needs to keep doing its thing
4. locktower needs to keep running

Reply:
The term `EntryTree` no longer exists in the implementation or book. It's now called `Blocktree` in the book and `DbLedger` in the implementation.  Also, it's a data structure and does not "run".

Reply:
@rob-solana 

1. That should be pulled out into `ClusterInfo`, or the network protocol layer.  That layer shouldn't need access to the bank.

2. Why?  Once this assumes the leader slot, it is using all of its compute for running the leader transactions.

3. Yep, block tree is still receiving data.

4. Possibly.  It can catch up after the leader slot is over.

Reply:
> The term `EntryTree` no longer exists in the implementation or book. It's now called `Blocktree` in the book and `DbLedger` in the implementation. Also, it's a data structure and does not "run".

@garious what do you call the code that adds blocks to blocktree at runtime?  That part is factored out of the loop.

Reply:
In the TPU, it used to be called WriteStage. I think that got absorbed by either BankingStage or BroadcastService.  In the TVU, I believe that was one of the responsibilities of the RetransmitStage.

Reply:
broadcast service has absorbed the role of writing the leader's data to db_ledger for now, might be nice to move it back out at some point...

Reply:
@garious @rob-walker I think the ledger write/retransmit can live outside of the tpu/tvu because the leader schedule persists for an epoch and is derived ahead of time from a root checkpoint.  So it can be factored out.

Reply:
Broadcast Service is outside (just after) the TPU. Feels like a nice place for that DbLedger write, since it's effectively broadcasting entries to itself, simply using DbLedger to alleviate the need to `sigverify` and `ReplayStage::process_entries()` in the TVU on data it already knows is valid.